File: ModuleCommon/src/main/java/com/qihuan/commonmodule/base/BaseMvpActivity.java
package com.qihuan.commonmodule.base;

import android.os.Bundle;
import android.support.annotation.Nullable;

/**
 * BaseMvpActivity
 *
 * @author qi
 * @date 2018/5/28
 */
public abstract class BaseMvpActivity<P extends BasePresenter> extends BaseActivity implements BaseView {

    private P mPresenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenter = initPresenter();
        if (mPresenter != null) {
            // noinspection unchecked
            mPresenter.attachView(this);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showError(String errorMsg) {

    }

    protected abstract P initPresenter();

}

File: module_common/src/main/java/com/qihuan/commonmodule/base/BaseMvpActivity.kt
package com.qihuan.commonmodule.base

import android.os.Bundle

/**
 * BaseMvpActivity
 *
 * @author qi
 * @date 2018/5/28
 */
abstract class BaseMvpActivity<in V : BaseView, out P : BasePresenter<V>> : BaseActivity(), BaseView {

    protected val mPresenter: P by lazy { initPresenter() }

    @Suppress("UNCHECKED_CAST")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenter.attachView(this as V)
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.detachView()
    }

    override fun showLoading() {

    }

    override fun hideLoading() {

    }

    override fun showError(errorMsg: String) {

    }

    protected abstract fun initPresenter(): P

}
